cmd/oci-sync/main.go

```diff
@@ -93,9 +95,35 @@ func main() {
 	failCount := 0
 	backoff := errorBackoff()
 
+	var authenticator authn.Authenticator
+	switch configsync.AuthType(*flAuth) {
+	case configsync.AuthNone:
+		authenticator = authn.Anonymous
+	case configsync.AuthGCPServiceAccount, configsync.AuthK8sServiceAccount, configsync.AuthGCENode:
+		// authenticator = google.NewEnvAuthenticator()
+		authenticator = &oci.CredentialAuthenticator{
```

Comment:
Keeping the authenticator outside of the for loop makes it impossible to auto-refresh the authenticator. For example, if the IAM policy bindings are changed, the authenticator will stay invalid unless the Pod is restated.

Reply:
The authenticator here is just a thin wrapper around the CachingCredentialProvider. So using the same authenticator instance allows caching the Credentials. And the new Credentials library handles caching the token as long as it's valid, with early refresh by default.

If we don't use the same authenticator, then we have to re-detect the credentials and re-request a new token every time, which involves one or two API calls, depending on the config.

If the IAM policy changes, does the client really need a new token? Or does it just check the bindings server-side when it makes a new API call?

Reply:
CachedTokenProvider: https://github.com/googleapis/google-cloud-go/blob/auth/v0.12.0/auth/auth.go#L282

`cachedTokenProvider.Token()` returns the cached token if it's non-nil, has a non-empty token value, and has not expired, with early refresh. Otherwise it gets a new token from the wrapped TokenProvider. Errors are not cached.

Using Fleet Workload Identity with GOOGLE_APPLICATION_CREDENTIALS set, `readCredentialsFileJSON` will return a `New3LOTokenProvider`, which builds a `NewCachedTokenProvider` wrapping a `tokenProvider3LO`, which fetches a access token & refresh token from the url configured in the json.

So you're right, it won't fetch a new token every time, but that's a feature, not a bug. That's why the caching is part of the core library now.

Reply:
I'm having trouble finding a source, but I'm 99% sure that Google IAM policy is not stored in the access token. I'm pretty sure the access token is an encrypted JWT that the server can decrypt with the public key of the auth server, but it should only include the scopes that the client requested, which allows the client to prove it has permission to use those scopes, signed by the auth server, but IAM policy must still be checked for those scopes by the server. So if the IAM policy changes, the client doesn't need a new token to pick up those changes.

The client should only need a new token when the token expires or the client wants to change the scopes it uses. Since clients rarely change scopes (maybe after an upgrade) and scopes rarely change on the server (maybe if services are added), the auth server grants permission to use those scopes for the duration of the token, usually 24hr.

Reply:
ACK. Thanks for looking it up. I recall seeing the permission denied error as a server response, so I think you're right, the IAM policy bindings should be checked by the server.